### PR TITLE
Remove the concepts of the regular key

### LINKED ISSUE #39: Remove the concept of "regular key"

### PR DESCRIPTION
**Pull request by somniumism**
To resolves #39 , I closed the @cubismic 's PR #132  and created a new PR.

I removed the concepts of the regular key.

### CODE REVIEW

**Comment by somniumism** [comment]
I have some problem that 22 tests are failed. I found the cause of the failures using the backtrace, and it is related to rlp encode. In scheme/scheme.rs, `initializae_accounts()`is followings:
The error or failure process can be schemed as follows:
1. `initialize_accounts()` in scheme/scheme.rs
2. `&account.rlp_bytes()`. In here, the type of account is `PodAccount` struct.
3. `s.drain() `
4. `assert!(self.is_finished());`
5. panicked

To sum up, failures of the 22 tests are occur from [is_finished() function](https://docs.rs/ethereum-rlp/0.2.3/rlp/struct.RlpStream.html#method.is_finished) in `rlp::RlpStream`. I'm trying to solve this cause, and if anyone knows a good way, please advise.

File: state/src/item/account.rs [hunk 1]
```diff
@@ -131,16 +104,15 @@ impl Encodable for Account {
         s.append(&PREFIX);
         s.append(&self.balance);
         s.append(&self.seq);
-        s.append(&self.regular_key);
```
Comment (sgkim126) [inline state/src/item/account.rs]: You must decrease the number of the list.

Comment (somniumism) [reply state/src/item/account.rs]: I modified it.

**Comment by somniumism** [comment]
I fixed the size of the list related to rlp encode, so I solved the problem that 22 tests are failed. However, new failure occurs and its cause is as follows:
```
---- impls::top_level::tests_state::ensure_cached stdout ----
thread 'impls::top_level::tests_state::ensure_cached' panicked at 'assertion failed: `(left == right)`
  left: `Ok(0xdb4046bb91a12a37cbfb0f09631aad96a97248423163eca791e19b430cc7fe4a)`,
 right: `Ok(0x9e14103f4537534c835b69916aa554cb7d2064f07a86d569949b971576893116)`', state/src/impls/top_level.rs:966:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

---- item::account::tests::new_account stdout ----
thread 'item::account::tests::new_account' panicked at 'assertion failed: `(left == right)`
  left: `"c3434580"`,
 right: `"c4434580c0"`', state/src/item/account.rs:159:9
```
Tests are here:
```rust
    fn ensure_cached() {
        let mut state = get_temp_state();
        let a = Address::default();
        state.get_account_mut(&a).unwrap();
        assert_eq!(Ok(H256::from("db4046bb91a12a37cbfb0f09631aad96a97248423163eca791e19b430cc7fe4a")), state.commit());
    }

    fn new_account() {
        let a = Account::new(69, 0);
        assert_eq!(a.rlp_bytes().to_hex(), "c4434580c0");
        assert_eq!(69, a.balance());
        assert_eq!(0, a.seq());
    }
```
This seems to be an error that occurs from that it is different from the hash value and the rlp value because the components of the rlp changed. To solve this problem, I should change values such as "c4434580c0" and "db4046bb91a12a37cbfb0f09631aad96a97248423163eca791e19b430cc7fe4a". 
@sgkim126 Can I change these values?

**Comment by majecty** [comment]
@somniumism Yes, since this PR changed fields in Account, you should update the hashes.

**Comment by somniumism** [comment]
It's all done. : ) @sgkim126 @majecty @HoOngEe Could you review this PR?

File: package-lock.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+  "lockfileVersion": 1
+}
```
Comment (majecty) [inline package-lock.json]: Please remove this file.

File: spec/JSON-RPC.md [hunk 3]
```diff
@@ -1402,7 +1330,7 @@ It returns null if the block number parameter is larger than the current best bl
 ### Returns
 `[number, number]` | `null`
 
-- The first item is the last block number that the term is closed. 
+- The first item is the last block number that the term is closed.
```
Comment (majecty) [inline spec/JSON-RPC.md]: Do not include unrelated code diff.

File: test/package-lock.json [hunk 4]
```diff
@@ -0,0 +1,3193 @@
+{
```
Comment (majecty) [inline test/package-lock.json]: Let's remove this file.

File: test/src/e2e/changeParams.test.ts [hunk 5]
```diff
@@ -178,7 +178,7 @@ describe("ChangeParams", function() {
             0x0200, // maxTextContentSize
             "tc", // networkID
             11, // minPayCost
-            10, // minSetRegularKeyCost
+            // 10, // minSetRegularKeyCost
```
Comment (majecty) [inline test/src/e2e/changeParams.test.ts]: Let's remove these lines.

File: test/src/e2e/termChange.test.ts [hunk 6]
```diff
@@ -57,7 +57,7 @@ describe("Term change", function() {
             0x0200, // maxTextContentSize
             "tc", // networkID
             10, // minPayCost
-            10, // minSetRegularKeyCost
+            // 10, // minSetRegularKeyCost
```
Comment (majecty) [inline test/src/e2e/termChange.test.ts]: ditto

File: types/src/transaction/action.rs [hunk 7]
```diff
@@ -18,15 +18,14 @@ use crate::errors::SyntaxError;
...
-use ckey::{recover, Address, NetworkId, Public, Signature};
+use ckey::{recover, Address, NetworkId, Signature};
 use primitives::{Bytes, H256};
 use rlp::{Decodable, DecoderError, Encodable, Rlp, RlpStream};
 
 #[derive(Clone, Copy)]
 #[repr(u8)]
 enum ActionTag {
     Pay = 0x02,
-    SetRegularKey = 0x03,
```
Comment (HoOngEe) [inline types/src/transaction/action.rs]: @somniumism How about pulling the ActionTag numbers one by one to fill from `0x03`?

Comment (somniumism) [reply types/src/transaction/action.rs]: You mean as follows?
```rust
enum ActionTag {
    Pay = 0x02,
    CreateShard = 0x03,
    SetShardOwners = 0x04,
    SetShardUsers = 0x05,
    ShardStore = 0x19,
    Custom = 0xFF,
}
```

Comment (somniumism) [reply types/src/transaction/action.rs]: or
```rust
enum ActionTag {
    Pay = 0x02,
    CreateShard = 0x03,
    SetShardOwners = 0x04,
    SetShardUsers = 0x05,
    ShardStore = 0x06,
    Custom = 0x07,
}
```

Comment (HoOngEe) [reply types/src/transaction/action.rs]: I prefer the above one. I guess `Custom` ActionTag was set to `0xFF` intentionally but I'm not sure about `ShardStore`

Comment (somniumism) [reply types/src/transaction/action.rs]: Thank you. I modified it. : )

**Comment by HoOngEe** [comment]
Check the e2e test and correct it, please.

**Comment by somniumism** [comment]
I checked the e2e test, but I found a problem I couldn't solve. I couldn't ascertain what caused this problem. I need someone's help.

```sh
    Sending invalid transactions over the limits (in action 6: SetShardUsers)
      1) shardId: 65536
Keep log file: /Users/somniumism/Project/PR/foundry/test/src/helper/../../../test/log/2020_02_12T12_44_24_938Z.61.log
      2) shardId: 100000
      - Users


  140 passing (32s)
  10 pending
  2 failing

  1) solo - 1 node
       Sending invalid transactions over the limits (in action 6: SetShardUsers)
         shardId: 65536:

      AssertionError: not similar
      + expected - actual

       {
         "code": -32009
      -  "data": "Custom(\"Unexpected action prefix\")"
      -  "message": "Invalid RLP."
      +  "data": "$containsWord(RlpIsTooBig)"
      +  "message": "$anything"
       }
      
      at Context.<anonymous> (src/e2e/verification.test.ts:290:34)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)

  2) solo - 1 node
       Sending invalid transactions over the limits (in action 6: SetShardUsers)
         shardId: 100000:

      AssertionError: not similar
      + expected - actual

       {
         "code": -32009
      -  "data": "Custom(\"Unexpected action prefix\")"
      -  "message": "Invalid RLP."
      +  "data": "$containsWord(RlpIsTooBig)"
      +  "message": "$anything"
       }
      
      at Context.<anonymous> (src/e2e/verification.test.ts:290:34)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

File: test/src/e2e/verification.test.ts [hunk 8]
```diff
@@ -123,14 +123,12 @@ describe("solo - 1 node", function() {
...
-            { actionType: 4, actionLength: 1 }, // CreateShard
-            { actionType: 4, actionLength: 3 },
-            { actionType: 5, actionLength: 2 }, // SetShardOwners
+            { actionType: 4, actionLength: 2 }, // SetShardOwners
+            { actionType: 4, actionLength: 4 },
+            { actionType: 5, actionLength: 2 }, // SetShardUsers
             { actionType: 5, actionLength: 4 },
-            { actionType: 6, actionLength: 2 }, // SetShardUsers
-            { actionType: 6, actionLength: 4 },
             { actionType: 0x19, actionLength: 3 }, // ShardStore
```
Comment (majecty) [inline test/src/e2e/verification.test.ts]: @somniumism You should not change these lines. These changes may cause the tests to fail.

Comment (somniumism) [reply test/src/e2e/verification.test.ts]: I fixed the cause of the test failure. However, why does this cause test failure? I changed this `verification.test.ts` because I changed `ActionTag` to new order:
original `ActionTag`
```rust
enum ActionTag {
    Pay = 0x02,
    CreateShard = 0x04,
    SetShardOwners = 0x05,
    SetShardUsers = 0x06,
    ShardStore = 0x19,
    Custom = 0xFF,
}
```
new order `ActionTag`:
enum ActionTag {
    Pay = 0x02,
    CreateShard = 0x03,
    SetShardOwners = 0x04,
    SetShardUsers = 0x05,
    ShardStore = 0x19,
    Custom = 0xFF,
}
```
So, I though I should change this `verification.test.ts`. I think it's right to modify it, but I don't understand why does this cause test failure. Could you explain why?

For the present, I modified to use the original ActionTag sequence to solve this error.

**Comment by somniumism** [comment]
I solved cause the tests to fail, and I checked that all e2e tests are clear. As @HoOngEe suggested, I could not make the following corrections: 
```rust
enum ActionTag {
    Pay = 0x02,
    CreateShard = 0x03,
    SetShardOwners = 0x04,
    SetShardUsers = 0x05,
    ShardStore = 0x19,
    Custom = 0xFF,
}
```
I didn't know the exact reason, but it's because this modification causes a test failure. I think it would be better to make this modification on other issue next time.

In conclusion, all modifications have been completed.

**Comment by HoOngEe** [comment]
OK, I checked. Let's resolve the conflicts.

**Comment by somniumism** [comment]
@HoOngEe I resolved the conflicts : )
